tests/test_builtin_traits.py

```diff
@@ -258,6 +263,180 @@ def test_no_terminator_op_with_is_terminator_op():
     op0.verify()
 
 
+@irdl_op_definition
+class IsSingleBlockImplicitTerminatorOp(IRDLOperation):
+    """
+    An operation that implements terminator to be used with an operation that uses the
+    SingleBlockImplicitTerminator trait.
+    """
+
+    name = "test.is_single_block_implicit_terminator"
+
+    # TODO fix circular reference
+    # traits = frozenset([HasParent(HasSingleBlockImplicitTerminatorOp), IsTerminator()])
+    # this is tracked by gh issue: https://github.com/xdslproject/xdsl/issues/1218
+    traits = frozenset([IsTerminator()])
+
+
+@irdl_op_definition
+class HasSingleBlockImplicitTerminatorOp(IRDLOperation):
+    """
+    An operation that expects a single-block region and an implicit terminator trait for
+    that block.
+    """
+
+    name = "test.has_single_block_implicit_terminator"
+
+    irdl_options = [AttrSizedRegionSegments()]
+
+    region: Region = region_def()
+    opt_region: OptRegion = opt_region_def()
+
+    traits = frozenset(
+        [SingleBlockImplicitTerminator(IsSingleBlockImplicitTerminatorOp)]
+    )
+
+    def __post_init__(self):
```

Comment:
I'm not sure this is good practice to have a __post_init__, as this may trigger unexpected behavior from the user when using `create` for instance.
I think it's fine for this test though (though I am a bit scared people might take examples from this test directly).

Reply:
Maybe add a comment there to specify that?

Reply:
Personally, I feel like this is exactly the kind of think we want to have in a post_init. Can we make it work with create as well? (Maybe open an issue on this, @math-fehr ?)

Reply:
Personally, I like putting that in `__post_init__`.
@math-fehr what would be the alternative place to call `ensure_terminator`?

Reply:
Okay, if you both feel like this is fine, then let's leave it!
I am just scared we end up in a situation where a rewrite doesn't work because `ensure_terminator` is called in an unexpected place.

I feel that if we abstract this nicely, and have a list of `post_init` calls, maybe we can have an option `disable_post_init` for an escape hatch?

Let's leave that for another time, and let's put an issue here instead, and keep the `__post_init__`.

Reply:
I'm in no rush to merge this in.
I've been further looking into this and it seems that the custom `__post_init__` in the `ImplicitBuilder` does pose a problem for this right now.

I was planning to update today, but I'll have a better idea tomorrow, so maybe we can take it to Zulip.

Reply:
For that there are two solutions: add an escape hatch to disable the implicit builder, or just use an implicit builder for the ensure terminator mechanism. I'd recommend trying the second one, might not be too bad.

Reply:
I prefer the latter and it's also similar to what MLIR does.
I'm not sure if we want the escape hatch mechanism on the `ensure_terminator` for now.
I'll finish up the addition of the trait in the dialects and open a PR stacked on this one; it might inform the discussion further.

Reply:
It seems this is also problematic with the `clone` in `core/ir.py`.
Moreover, I'm not sure we want to bring the builders to the core (which is where traits belong), so I'm reconsidering my inclination toward the builders.

I'll have a look into an "escape hatch" for them.

Reply:
IMO, this is starting to become a case of PR creep and I would follow Mathieu's suggestion here and upstream this with an open issue/a quick fix. We will not implement the solution in this PR anyways, so we might as well actually solve it later on.

As for the different decisions, I don't think I have the full technical insight needed to give a strong opinion. I am personally against taking the builders into core (if I understood the suggestion properly), but if you guys say this is the most reasonable way of doing it, then I will trust you on that.

Reply:
That is the point of this discussion IMHO.
I've opened https://github.com/xdslproject/xdsl/pull/1248 but I'm not sure it exemplifies the issues just by looking  at it (hence the discussion here and on Zulip :-)